addons/format_lame.c

```diff
@@ -0,0 +1,559 @@
+	/* Sample rate */
+	lame_set_in_samplerate(p->lgfp, p->sample_rate);
+	/* Bit rate, e.g.:
+	* 16kbps for 8000Hz
+	* 32kbps for 16000Hz
+	* 64kbps for 32000Hz
+	* 96kbsp for 48000Hz
+	*/
+	lame_set_brate(p->lgfp, p->sample_rate / 500);
+	/* Mono */
+	lame_set_mode(p->lgfp, 3);
+	/* Medium quality */
+	lame_set_quality(p->lgfp, 5);
+	return lame_init_params(p->lgfp);
+}
+
+static int mp3lame_encoder_deinit(struct ast_filestream *fs)
+{
+	int res;
+	int wres;
+	struct mp3lame_private *p = fs->_private;
+	unsigned char last_encoder_buffer[p->sample_rate];
+
+	res = lame_encode_flush(p->lgfp, last_encoder_buffer, sizeof(last_encoder_buffer));
+	if (res < 0) {
+		ast_log(LOG_WARNING, "LAME encode last returned error %d", res);
+		lame_close(p->lgfp);
+		return -1;
+	} else if (res > 0) {
+		wres = fwrite(last_encoder_buffer, 1, res, fs->f);
+		if (wres != res) {
+			ast_log(LOG_WARNING, "Bad write (%d/%d): %s", wres, res, strerror(errno));
+			lame_close(p->lgfp);
+			return -1;
+		} else {
+			ast_debug(3, "LAME encode flushed %d bytes", wres);
+		}
+	}
+
+	return lame_close(p->lgfp);
+}
+
+static int mp3lame_decoder_init(struct ast_filestream *fs)
+{
+	int size;
+	struct mp3lame_private *p = fs->_private;
+
+	p->hgfp = hip_decode_init();
+	if (!p->hgfp) {
+		return -1;
+	}
+	/* Get filesize in bytes */
+	fseek(fs->f, 0, SEEK_END);
+	size = ftell(fs->f);
+	fseek(fs->f, 0, SEEK_SET);
+	return size;
+}
+
+static int mp3lame_decoder_deinit(struct mp3lame_private *p)
+{
+	return hip_decode_exit(p->hgfp);
+}
+
+static int mp3lame_hip_decode(hip_global_flags * handle, int input_size, unsigned char * input, short int * output, int sample_rate)
+{
+	int samples;
+	mp3data_struct headers;
+
+	samples = hip_decode_headers(handle, input, input_size, output, NULL, &headers);
+
+	if (headers.header_parsed == 1) {
+		ast_debug(3, "LAME decoder found MP3 headers:\nchannels=%d, samplerate=%d, bitrate=%d, framesize=%d, mode=[%d:%d]",
+			headers.stereo,
+			headers.samplerate,
+			headers.bitrate,
+			headers.framesize,
+			headers.mode,
+			headers.mode_ext
+		);
+		if (headers.stereo < 1) {
+			ast_debug(3, "LAME decoder no audio channels");
+			return -1;
+		} else if (headers.stereo > 1) {
+			ast_log(LOG_ERROR, "LAME decoder invalid number of channels: %d, only mono is acceptable", headers.stereo);
+			return -2;
+		}
+		if (headers.samplerate != sample_rate) {
+			ast_log(LOG_ERROR, "LAME decoder invalid sampling rate: %d, expected %d", headers.samplerate, sample_rate);
+			return -3;
+		}
+	}
+
+	return samples;
+}
+
+static int mp3lame_file_read(struct ast_filestream *s, int sample_rate)
+{
+	int rres;
+	struct mp3lame_private *p = s->_private;
+	unsigned char dinput[p->total_bytes_compressed];
+
+	rres = fread(dinput, 1, p->total_bytes_compressed, s->f);
+	if (rres != p->total_bytes_compressed) {
+		ast_log(LOG_ERROR, "Short read (%d/%d)", rres, p->total_bytes_compressed);
+		return -1;
+	}
+
+	ast_debug(3, "LAME decoder input: %d bytes", rres);
```

Comment:
Maybe increase the debug level of these messages? They seem a bit low for something that would print on every read callback